### PR TITLE
Store API

### LINKED ISSUE #10: Store API
Notes on implementation:
- another bug has been found in lmdb. [bug report here](https://bugs.openldap.org/show_bug.cgi?id=9920). I think I fixed it.
- About the sizes of entries in LMDB, here is what I understood so far:
  * our objectIds are represented as 33 bytes (1 for the enum, and 32 for the blake3 hash)
  * An empty Object is taking 6 bytes already (without any content, deps, expiry or children)
  * For LMDB, what matters is the total combined size of the key and the value.
  * The tests that I conduced bring me those empirical sizes for the value (knowing that the key is 33 bytes): 
    * from 1 to 1977 bytes, the value stays on one memory page
    * from 1988 to 4072, an overflow page is added (so we already have 2 pages used to store the value)
    * above 4072, a new page is added for each 4096 bytes chunk.
    * So the formula can be kept simple: our values will use optimal space when following the formula : 4072 + n x 4096 bytes.
    * there is no practical limit to the upper bound of the size of a value. But we have already decided to stay below 2MB.

### PR DESCRIPTION
**Pull request by nikoPLP**
don't forget to do a 
```
cargo update
cargo clean
```

### CODE REVIEW

File: lofire/src/types.rs [hunk 1]
```diff
@@ -52,6 +52,8 @@ pub enum Signature {
 /// Timestamp: absolute time in minutes since 2022-02-22 22:22 UTC
 pub type Timestamp = u32;
 
+pub const EPOCH_LOFIRE_IN_UNIX_TIMEPSTAMP: u64 = 1645568520;
```
Comment (tg-x) [inline lofire/src/types.rs]: there's a typo, and it's too long, I suggest instead `EPOCH_AS_UNIX_TIMESTAMP`

File: lofire-repo/src/store.rs [hunk 2]
```diff
@@ -1,10 +1,437 @@
...
+        Store {
+            environment: shared_rkv.clone(),
+            main_store,
+            meta_store,
+            expiry_store,
+            recently_used_store,
+        }
+    }
+
+    fn remove_from_LRU(
```
Comment (tg-x) [inline lofire-repo/src/store.rs]: i'd keep fn names all lowercase

File: lofire-repo/src/store.rs [hunk 3]
```diff
@@ -1,10 +1,437 @@
...
+        &self,
+        writer: &mut Writer<LmdbRwTransaction>,
+        obj_id_ser: &Vec<u8>,
+        time: &Timestamp,
+    ) -> Result<(), StoreError> {
+        self.recently_used_store
+            .delete(writer, *time, &Value::Blob(obj_id_ser.as_slice()))
+    }
+
+    fn add_to_LRU(
```
Comment (tg-x) [inline lofire-repo/src/store.rs]: same here

File: lofire-repo/src/store.rs [hunk 4]
```diff
@@ -1,10 +1,437 @@
...
+        flag.set(WriteFlags::APPEND_DUP, true);
+        self.recently_used_store.put_with_flags(
+            writer,
+            *time,
+            &Value::Blob(obj_id_ser.as_slice()),
+            flag,
+        )
+    }
+
+    pub fn pin(&self, object_id: &ObjectId, add: bool) -> Result<(), Error> {
```
Comment (tg-x) [inline lofire-repo/src/store.rs]: i'd rename this to `set_pin()` for clarity (otherwise would be expecting a `pin()` & `unpin()` pair)

Comment (tg-x) [reply lofire-repo/src/store.rs]: also consider adding `is_pinned(id)`

File: lofire-repo/src/store.rs [hunk 5]
```diff
@@ -1,10 +1,437 @@
...
+
+// TODO: versioning V0
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+struct ObjectMeta {
+    pub pin: bool,
+    pub last_used: Timestamp,
+    pub synced: bool,
+}
+
+impl Store {
```
Comment (tg-x) [inline lofire-repo/src/store.rs:43]: let's add doc a comment for each function

**Review by tg-x — CHANGES_REQUESTED** [review]

File: lofire-repo/src/store.rs [hunk 6]
```diff
@@ -1,10 +1,437 @@
...
+    LmdbDatabase, LmdbDatabaseFlags, LmdbEnvironment, LmdbRwTransaction, LmdbWriteFlags,
+};
+use rkv::{
+    Manager, MultiIntegerStore, Rkv, SingleStore, StoreError, StoreError::DataError, StoreOptions,
+    Value, WriteFlags, Writer,
+};
+use serde_bare::error::Error;
+
+pub struct Store {
+    // the main store where all the repo objects are stored
```
Comment (tg-x) [inline lofire-repo/src/store.rs:23]: use `/// doc comments` here

Comment (tg-x) [inline lofire-repo/src/store.rs]: use `/// doc comments` here

**Review by tg-x — CHANGES_REQUESTED** [review]

File: lofire-repo/src/store.rs (hunk 4 above)
Comment (nikoPLP) [reply lofire-repo/src/store.rs]: is_pinned i will implement later. if we need it at some point

**Review by tg-x — APPROVED** [review]
LGTM

**Pull request merged**